File: scripts/atomic_ledger.py
#!/usr/bin/env python3
"""Reproduce the isotope-count table and circular two-end energy/action ledger.

Only standard-library Python is required. Run from any directory. Data outputs
are placed beside this script. The isotope list is inherited exhibit data;
its proposed physical partition is not inferred by these arithmetic checks.
"""
import ast
import csv
import json
import re
from decimal import Decimal as D, localcontext
from pathlib import Path

HERE = Path(__file__).resolve().parent
SOURCE = HERE.parent / 'poams-periodic-table.html'
INPUTS = json.loads((HERE / 'coulomb_calibrated_inputs.json').read_text())
PI = D('3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628')
CHECKS = []


def check(label, condition):
    if not condition:
        raise AssertionError(label)
    CHECKS.append(label)


def close(label, actual, expected):
    scale = max(abs(actual), abs(expected), D('1e-100'))
    check(label, abs(actual - expected) / scale < D('1e-65'))


def data():
    raw = re.search(r'const RAW_ELEMENTS = (\[.*?\n        \]);', SOURCE.read_text(), re.S)
    if not raw:
        raise ValueError('Cannot locate inherited element data')
    records = ast.literal_eval(raw[1])
    check('118 distinct sequential element labels', [r[0] for r in records] == list(range(1, 119)))
    rows = []
    for Z, symbol, name, _kind, _mass, A, isotopes in records:
        available = list(dict.fromkeys([A, *(int(re.search(r'-(\d+)', s)[1]) for s in isotopes)]))
        check(f'{symbol}: valid isotope labels', all(a >= Z and isinstance(a, int) for a in available))
        rows.append(dict(Z=Z, symbol=symbol, name=name, A=A, depth=A-Z, pattern=Z,
                         ratio=str(D(A-Z)/Z), depth_fraction=str(D(A-Z)/A),
                         available_isotopes=available))
    check('carbon isotope example', [a-6 for a in rows[5]['available_isotopes']] == [6,7,8])
    check('hydrogen zero surplus is a count result', rows[0]['depth'] == 0)
    return rows


def circular(model, n):
    value = lambda key: D(INPUTS['inputs'][key]['value'])
    h, c, alpha = value('h'), value('c'), value('alpha')
    m1 = value('m_light')
    m2 = value('m_heavy') if model == 'pair' else m1
    M = m1+m2
    mu = m1*m2/M
    hbar = h/(2*PI)
    L = n*hbar
    K = alpha*hbar*c
    # Solve dE/dr=0 at fixed angular momentum, then calculate the endpoints
    # directly. Binding is computed from the sum, not stipulated as T.
    r = L*L/(mu*K)
    omega = L/(mu*r*r)
    r1, r2 = m2*r/M, m1*r/M
    T1, T2 = m1*(omega*r1)**2/2, m2*(omega*r2)**2/2
    L1, L2 = m1*r1*r1*omega, m2*r2*r2*omega
    U = -K/r
    B = -(T1+T2+U)
    f = omega/(2*PI)
    nu = B/h
    S1, S2 = 2*PI*L1, 2*PI*L2
    label = f'{model}, n={n}: '
    close(label+'stationarity', L*L/(mu*r**3), K/r**2)
    check(label+'positive radial curvature', 3*L*L/(mu*r**4)-2*K/r**3 > 0)
    close(label+'barycentre', m1*r1, m2*r2)
    close(label+'relative radius sum', r1+r2, r)
    close(label+'angular-momentum sum', L1+L2, L)
    close(label+'virial and one joint energy', 2*(T1+T2), -U)
    close(label+'binding expression', B, mu*c*c*alpha*alpha/(2*n*n))
    close(label+'rate distinction', nu/f, D(n)/2)
    close(label+'action sum', S1+S2, n*h)
    close(label+'energy share', T1/(T1+T2), m2/M)
    close(label+'angular-momentum share', L1/L, m2/M)
    close(label+'radius/reference ratio', r/(hbar/(mu*c)), n*n/alpha)
    close(label+'L omega is twice kinetic energy', L*omega, 2*(T1+T2))
    if model == 'equal':
        close(label+'equal endpoint split', T1, T2)
    return dict(model=model,n=n,mu=mu,K=K,radius=r,omega=omega,
                binding=B,bindingEV=B/value('e_SI'),bindingRate=nu,f=f,
                r1=r1,r2=r2,T1=T1,T2=T2,L=L,L1=L1,L2=L2,S1=S1,S2=S2,
                U=U,share1=m2/M,share2=m1/M)


def serialize(value):
    if isinstance(value, D):
        return format(value, '.30g')
    raise TypeError(type(value).__name__)


def main():
    with localcontext() as ctx:
        ctx.prec = 80
        rows = data()
        cases = [circular(model,n) for model in ('pair','equal') for n in range(1,7)]
        # Scaling checks compare separately solved cases, not a display label.
        for start in (0,6):
            base = cases[start]
            for q in cases[start+1:start+6]:
                n = q['n']
                close(f"{q['model']} n={n}: radius progression", q['radius']/base['radius'], D(n*n))
                close(f"{q['model']} n={n}: rate progression", base['f']/q['f'], D(n**3))
        result = dict(
            title='Atomic count and calibrated two-ended ledger', date='2026-09-15',
            inputs=INPUTS['inputs'],
            used_inputs=['alpha','c','h','e_SI','m_light','m_heavy'],
            source=INPUTS['source'],
            precision='80 working decimal digits, 30 printed; input uncertainty limits physical accuracy',
            assumptions=[
                'Counts are inherited neutral-isotope assignments, not measured regional angular momenta.',
                'Integer mass-number labels are not endpoint mass calibrations.',
                'E=L^2/(2 mu r^2)-K/r, K=alpha hbar c, L=n hbar; nonrelativistic circular scale model.',
                'The pair endpoints are not identified as a many-entry atom\'s core and envelope.',
                'Intrinsic spin, many-entry state coupling and region sizes are not calculated.',
                'The action here is angular action per circuit, not the full time-integrated Lagrangian action.'
            ],
            representatives=rows, pair_cases=cases,
            checks_passed=len(CHECKS), checks=CHECKS)
        (HERE/'atomic_ledger_results.json').write_text(json.dumps(result,default=serialize,indent=2)+'\n')
        with (HERE/'atomic_count_partitions.csv').open('w',newline='') as handle:
            writer=csv.DictWriter(handle,fieldnames=['Z','symbol','name','A','depth','pattern','ratio','depth_fraction'])
            writer.writeheader()
            writer.writerows({key:row[key] for key in writer.fieldnames} for row in rows)
        print(json.dumps(dict(checks_passed=len(CHECKS),representatives=len(rows),cases=len(cases),
                              fundamental=cases[0]),default=serialize,indent=2))


if __name__ == '__main__':
    main()

File: scripts/atomic_ledger.js
/* Atomic ledger: inherited isotope labels and a calibrated circular pair.
   The two computations are deliberately separate: a mass-number label is
   neither a measured endpoint mass nor a physical core-radius prescription. */
(function (root) {
  'use strict';
  const calibrated = typeof module === 'object' && module.exports
    ? require('./coulomb_calibrated.js') : root.POAMSCoulomb;
  function counts(Z, A) {
    if (!Number.isInteger(Z) || !Number.isInteger(A) || Z < 1 || A < Z) {
      throw new RangeError('The count ledger requires integer A ≥ Z ≥ 1');
    }
    return {Z, A, depth:A-Z, pattern:Z, ratio:(A-Z)/Z, depthFraction:(A-Z)/A};
  }
  function pair(model, n) {
    if (!['pair','equal'].includes(model)) throw new RangeError('Choose finite or equal endpoints');
    const q=calibrated.calculate(model,n,1,2), k=calibrated.constants;
    const m1=k.lightMass, m2=model==='pair'?k.heavyMass:k.lightMass;
    const M=m1+m2, hbar=k.h/(2*Math.PI), L=n*hbar;
    const share1=m2/M, share2=m1/M;
    const r1=share1*q.radius, r2=share2*q.radius;
    // Direct mechanical sums, not definitions from the displayed shares.
    const T1=m1*(q.omega*r1)**2/2, T2=m2*(q.omega*r2)**2/2;
    const L1=m1*r1*r1*q.omega, L2=m2*r2*r2*q.omega;
    return {...q,m1,m2,M,L,share1,share2,r1,r2,T1,T2,L1,L2,
      U:-q.K/q.radius, f:q.omega/(2*Math.PI), bindingRate:q.binding/k.h,
      S1:2*Math.PI*L1,S2:2*Math.PI*L2};
  }
  function mount(elements) {
    const $=id=>document.getElementById(id);
    if (!$('atomic-ledger')) return;
    const fmt=(x,d=7)=>x.toPrecision(d).replace(/(?:\.0+|(\.\d*?)0+)(e|$)/,'$1$2');
    const option=(value,label)=>new Option(label,String(value));
    elements.forEach(e=>$('al-element').add(option(e.n,`${e.n} · ${e.name} (${e.symbol})`)));
    const body=$('al-all-counts');
    elements.forEach(e=>{
      const q=counts(e.n,e.massNumber), row=body.insertRow();
      [`${e.symbol}-${q.A}`,q.Z,q.depth,q.pattern,fmt(q.ratio,5)].forEach(value=>{
        const cell=row.insertCell();cell.textContent=String(value);
      });
    });
    function isotopeOptions() {
      const e=elements.find(e=>e.n===Number($('al-element').value));
      const values=[...new Set([e.massNumber,...e.isotopes.map(s=>Number(s.match(/-(\d+)/)[1]))])];
      $('al-isotope').replaceChildren(...values.map(A=>option(A,`${e.symbol}-${A}`)));
      renderCounts();
    }
    function renderCounts() {
      const q=counts(Number($('al-element').value),Number($('al-isotope').value));
      $('al-count-summary').textContent=`${q.depth} depth : ${q.pattern} pattern — ratio ${fmt(q.ratio)} : 1`;
      $('al-count-fractions').textContent=`${fmt(100*q.depthFraction,5)}% depth and ${fmt(100*(1-q.depthFraction),5)}% pattern, of ${q.A} assigned entries.`;
      $('al-depth-bar').style.width=`${100*q.depthFraction}%`;
      $('al-pattern-bar').style.width=`${100*(1-q.depthFraction)}%`;
      $('al-count-note').textContent=q.depth===0
        ? 'Zero surplus depth entries does not mean zero intrinsic turning or no compact structure.'
        : 'These are assigned counts, not measured regional masses, energy fractions or resultant angular momenta.';
    }
    function renderPair() {
      const q=pair($('al-mass').value,Number($('al-n').value)), k=calibrated.constants;
      $('al-n-label').textContent=String(q.n);
      $('al-radius').textContent=fmt(q.radius*1e12)+' pm';
      $('al-binding-rate').textContent=fmt(q.bindingRate/1e15)+' PHz';
      $('al-circulation').textContent=fmt(q.f/1e15)+' PHz';
      $('al-binding-units').textContent=fmt(q.binding)+' J = '+fmt(q.bindingEV)+' eV';
      $('al-virial').textContent=`T/h = +${fmt((q.T1+q.T2)/k.h/1e15)} PHz; U/h = −${fmt(-q.U/k.h/1e15)} PHz; E/h = −${fmt(q.bindingRate/1e15)} PHz.`;
      $('al-rate-ratio').textContent=`νB / fcirc = n/2 = ${q.n/2}. A binding-rate equivalent is not the circulation frequency.`;
      $('al-endpoint-label').textContent=q.model==='pair'?'Finite light/heavy calibration':'Equal-light-end diagnostic';
      [['1',q.r1,q.T1,q.L1,q.S1,q.share1],['2',q.r2,q.T2,q.L2,q.S2,q.share2]].forEach(([id,r,T,L,S,share])=>{
        $(`al-r${id}`).textContent=fmt(r*1e12)+' pm';
        $(`al-t${id}`).textContent=fmt(T/k.h/1e15)+' PHz';
        $(`al-j${id}`).textContent=fmt(L/(k.h/(2*Math.PI)))+' ℏ';
        $(`al-s${id}`).textContent=fmt(S/k.h)+' h';
        $(`al-share${id}`).textContent=fmt(share*100)+'%';
      });
      $('al-pair-total').textContent=`One pair: L = ${q.n} ℏ; angular action per full circuit = ${q.n} h; r/[ℏ/(μc)] = ${fmt(q.radiusRatio,10)}.`;
    }
    $('al-element').value='6';
    $('al-element').addEventListener('change',isotopeOptions);
    $('al-isotope').addEventListener('change',renderCounts);
    ['al-mass','al-n'].forEach(id=>$(id).addEventListener('input',renderPair));
    isotopeOptions();renderPair();
  }
  const api=Object.freeze({counts,pair,mount});
  root.POAMSAtomicLedger=api;
  if(typeof module==='object' && module.exports) module.exports=api;
})(typeof globalThis!=='undefined'?globalThis:this);
